src/L2/UpgradeableRegistrarController.sol

```diff
@@ -0,0 +1,686 @@
+    error ResolverRequiredWhenDataSupplied();
+
+    /// @notice Thrown when a `discountedRegister` claim tries to access an inactive discount.
+    ///
+    /// @param key The discount key that is inactive.
+    error InactiveDiscount(bytes32 key);
+
+    /// @notice Thrown when the payment received is less than the price.
+    error InsufficientValue();
+
+    /// @notice Thrown when the specified discount's validator does not accept the discount for the sender.
+    ///
+    /// @param key The discount being accessed.
+    /// @param data The associated `validationData`.
+    error InvalidDiscount(bytes32 key, bytes data);
+
+    /// @notice Thrown when the discount amount is 0.
+    ///
+    /// @param key The discount being set.
+    error InvalidDiscountAmount(bytes32 key);
+
+    /// @notice Thrown when the payment receiver is being set to address(0).
+    error InvalidPaymentReceiver();
+
+    /// @notice Thrown when the discount validator is being set to address(0).
+    ///
+    /// @param key The discount being set.
+    /// @param validator The address of the validator being set.
+    error InvalidValidator(bytes32 key, address validator);
+
+    /// @notice Thrown when a refund transfer is unsuccessful.
+    error TransferFailed();
+
+    /*´:°•.°+.*•´.*:˚.°*.˚•´.°:°•.°•.*•´.*:˚.°*.˚•´.°:°•.°+.*•´.*:*/
+    /*                          EVENTS                            */
+    /*.•°:°.´+˚.*°.˚:*.´•*.+°.•°:´*.´•*.•°.•°:°.´:•˚°.*°.˚:*.´+°.•*/
+
+    /// @notice Emitted when a discount is set or updated.
+    ///
+    /// @param discountKey The unique identifier key for the discount.
+    /// @param details The DiscountDetails struct stored for this key.
+    event DiscountUpdated(bytes32 indexed discountKey, DiscountDetails details);
+
+    /// @notice Emitted when an ETH payment was processed successfully.
+    ///
+    /// @param payee Address that sent the ETH.
+    /// @param price Value that was paid.
+    event ETHPaymentProcessed(address indexed payee, uint256 price);
+
+    /// @notice Emitted when a name was registered.
+    ///
+    /// @param name The name that was registered.
+    /// @param label The hashed label of the name.
+    /// @param owner The owner of the name that was registered.
+    /// @param expires The date when the registration expires.
+    event NameRegistered(string name, bytes32 indexed label, address indexed owner, uint256 expires);
+
+    /// @notice Emitted when a name is renewed.
+    ///
+    /// @param name The name that was renewed.
+    /// @param label The hashed label of the name.
+    /// @param expires The date that the renewed name expires.
+    event NameRenewed(string name, bytes32 indexed label, uint256 expires);
+
+    /// @notice Emitted when the payment receiver is updated.
+    ///
+    /// @param newPaymentReceiver The address of the new payment receiver.
+    event PaymentReceiverUpdated(address newPaymentReceiver);
+
+    /// @notice Emitted when the price oracle is updated.
```

Comment:
nit: the "price oracle" is also referred to as the "pricing oracle" and the "prices oracle" at various places in this file, might be good to standardize